Source/Tools/LinuxEmulation/LinuxSyscalls/x32/Memory.cpp

```diff
@@ -24,32 +26,54 @@ namespace FEX::HLE::x32 {
 void* x32SyscallHandler::GuestMmap(FEXCore::Core::InternalThreadState* Thread, void* addr, size_t length, int prot, int flags, int fd, off_t offset) {
   LOGMAN_THROW_A_FMT((length >> 32) == 0, "values must fit to 32 bits");
 
-  auto Result = (uint64_t)GetAllocator()->Mmap((void*)addr, length, prot, flags, fd, offset);
+  uint64_t Result {};
+  size_t Size = FEXCore::AlignUp(length, FEXCore::Utils::FEX_PAGE_SIZE);
 
-  LOGMAN_THROW_A_FMT((Result >> 32) == 0 || (Result >> 32) == 0xFFFFFFFF, "values must fit to 32 bits");
+  if (flags & MAP_SHARED) {
+    CTX->MarkMemoryShared(Thread);
+  }
+
+  {
+    // NOTE: Frontend calls this with a nullptr Thread during initialization, but
+    //       providing this code with a valid Thread object earlier would allow
+    //       us to be more optimal by using GuardSignalDeferringSection instead
+    auto lk = FEXCore::GuardSignalDeferringSectionWithFallback(VMATracking.Mutex, Thread);
+
+    Result =
+      (uint64_t) static_cast<FEX::HLE::x32::x32SyscallHandler*>(FEX::HLE::_SyscallHandler)->GetAllocator()->Mmap((void*)addr, length, prot, flags, fd, offset);
+
+    if (FEX::HLE::HasSyscallError(Result)) {
+      return reinterpret_cast<void*>(Result);
+    }
+
+    LOGMAN_THROW_A_FMT((Result >> 32) == 0 || (Result >> 32) == 0xFFFFFFFF, "values must fit to 32 bits");
 
-  if (!FEX::HLE::HasSyscallError(Result)) {
     FEX::HLE::_SyscallHandler->TrackMmap(Thread, Result, length, prot, flags, fd, offset);
-    return (void*)Result;
-  } else {
-    errno = -Result;
```

Comment:
What happened to this errno write? Not seeing any obvious match in the new version.